src/react/hooks/useQuery.ts

```diff
@@ -1,16 +1,355 @@
-import { DocumentNode } from 'graphql';
-import { TypedDocumentNode } from '@graphql-typed-document-node/core';
-
-import { QueryHookOptions, QueryResult } from '../types/types';
-import { useBaseQuery } from './utils/useBaseQuery';
+import { useContext, useEffect, useMemo, useRef, useState } from 'react';
+import { equal } from '@wry/equality';
 import { OperationVariables } from '../../core';
+import { getApolloContext } from '../context';
+import { ApolloError } from '../../errors';
+import {
+  ApolloQueryResult,
+  NetworkStatus,
+  ObservableQuery,
+  DocumentNode,
+  TypedDocumentNode,
+  WatchQueryOptions,
+} from '../../core';
+import {
+  QueryHookOptions,
+  QueryResult,
+} from '../types/types';
+
+import { DocumentType, verifyDocumentType } from '../parser';
+import { useApolloClient } from './useApolloClient';
 
-export function useQuery<TData = any, TVariables = OperationVariables>(
+export function useQuery<
+  TData = any,
+  TVariables = OperationVariables,
+>(
   query: DocumentNode | TypedDocumentNode<TData, TVariables>,
-  options?: QueryHookOptions<TData, TVariables>
-) {
-  return useBaseQuery<TData, TVariables>(query, options, false) as QueryResult<
-    TData,
-    TVariables
-  >;
+  options?: QueryHookOptions<TData, TVariables>,
+): QueryResult<TData> {
+  const context = useContext(getApolloContext());
+  const client = useApolloClient(options?.client);
+  verifyDocumentType(query, DocumentType.Query);
+  const [obsQuery, setObsQuery] = useState(() => {
+    const watchQueryOptions = createWatchQueryOptions(query, options);
+    // See if there is an existing observable that was used to fetch the same
+    // data and if so, use it instead since it will contain the proper queryId
+    // to fetch the result set. This is used during SSR.
+    let obsQuery: ObservableQuery<TData, TVariables> | null = null;
+    if (context.renderPromises) {
+      obsQuery = context.renderPromises.getSSRObservable(watchQueryOptions);
+    }
+
+    if (!obsQuery) {
+      // Is it safe (StrictMode/memory-wise) to call client.watchQuery here?
+      obsQuery = client.watchQuery(watchQueryOptions);
+      if (context.renderPromises) {
+        context.renderPromises.registerSSRObservable(
+          obsQuery,
+          watchQueryOptions,
+        );
+      }
+    }
+
+    if (
+      context.renderPromises &&
+      options?.ssr !== false &&
+      !options?.skip &&
+      obsQuery.getCurrentResult().loading
+    ) {
+      // TODO: This is a legacy API which could probably be cleaned up
+      context.renderPromises.addQueryPromise(
+        {
+          // The only options which seem to actually be used by the
+          // RenderPromises class are query and variables.
+          getOptions: () => createWatchQueryOptions(query, options),
```

Comment:
```suggestion
          getOptions: () => watchQueryOptions,
```
Do we need to recreate these options?

Reply:
`getOptions()` is called in a `useState` callback, and right now it only seems to have the initial hook options. I can’t reference the ref which holds a copy of options and such because it doesn’t exist yet. It’s a conundrum.

Reply:
Okay, good point about this code being SSR-specific. Suggestion withdrawn! 😄